### PR TITLE
Incorporating the availability of screen and depth textures for the GLES3 backend

### PR DESCRIPTION
**Pull request by the-brickster**
The PR is for the implementation of SCREEN_TEXTURE and DEPTH_TEXTURE for the GLES3 backend.

The link for the testbed for this PR can be found here: https://github.com/the-brickster/Godot4Testbed/tree/master

### CODE REVIEW

File: SConstruct
Comment (BastiaanOlij) [inline SConstruct:176]: These seem like changes you made just for your own debugging, not related to the GLES3 changes :)

File: drivers/gles3/storage/material_storage.cpp
Comment (BastiaanOlij) [inline drivers/gles3/storage/material_storage.cpp:1566]: This looks wrong, `projection_matrix` and `inv_projection_matrix` are calculated in the shader as they are assigned differently for multiview. This change will break multiview.

File: drivers/gles3/storage/texture_storage.cpp
Comment (BastiaanOlij) [inline drivers/gles3/storage/texture_storage.cpp:1577]: This needs multiview support, but we can add that in later. In theory it should be enough to switch between 2D and 3D textures if `use_multiview` is true and use `rt->viewcount` as the z component.

File: drivers/gles3/rasterizer_scene_gles3.cpp
Comment (BastiaanOlij) [inline drivers/gles3/rasterizer_scene_gles3.cpp:1993]: This needs multiview support, but we can add that in later. In theory it should be enough to set the z component to `view_count` if multiview is used

**Review by BastiaanOlij — CHANGES_REQUESTED** [review]
Generally looks good, just need to remove the changes to the build files and undo the break in projection matrix values.

If multiview could be taken along that would be a bonus, but we can add that later as well.

File: drivers/gles3/rasterizer_gles3.cpp
Comment (akien-mga) [inline drivers/gles3/rasterizer_gles3.cpp:297]: Please make sure not to include unnecessary changes in files with no functional change.

File: drivers/gles3/storage/texture_storage.h
Comment (akien-mga) [inline drivers/gles3/storage/texture_storage.h:324]: One line too many.

**Comment by akien-mga** [comment]
This seems to go directly against #70967...?

**Comment by BastiaanOlij** [comment]
> This seems to go directly against #70967...?

Ah I overlooked that, most of this PR is still valid but it indeed reintroduces the renames.

So yeah, @the-brickster we don't need to re-add `actions.renames["SCREEN_TEXTURE"]` and related things. There is a new way of doing this by adding a uniform with a hint in your shader.

So in your test project, in your shader simply add:
```
uniform sampler2D my_screen_texture : hint_screen_texture; 
```

Have a look at how things have changed in the Vulkan renderer if you're not sure. The PR Akien mentions should provide you all the further background.

**Comment by the-brickster** [comment]
Ahh ok, so that's why I needed to change it in my test project. I will get the required changes within the next day.

File: SConstruct
Comment (the-brickster) [reply SConstruct:176]: Will get this fixed in next commit

**Comment by the-brickster** [comment]
> This seems to go directly against #70967...?

I have integrated those changes into my PR

**Comment by the-brickster** [comment]
> @the-brickster I don't get it... How can this PR be "fixed" if the whole concept of `SCREEN_TEXTURE` and `DEPTH_TEXTURE` has been scrapped in #70967?
It hasn't been scrapped per se, it's just been replaced in favor of using `hint_depth_texture` and `hint_screen_texture`, you can confirm this by using the following test project I use:
https://github.com/the-brickster/Godot4Testbed/tree/master
Within the scene splat_test.tscn

**Comment by aaronfranke** [comment]
@the-brickster To avoid confusion, you can edit the title and description of the PR. It would also be a good idea to include a link to the test project you're using in the PR description.

**Comment by the-brickster** [comment]
@aaronfranke @Riteo title changed and testbed project added

**Comment by akien-mga** [comment]
@clayjohn needs to do a more in depth review of the implementation.

But you could already squash the commits, which would be needed before merging. See [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase).

BTW for future PRs, we recommend using a dedicated branch on your fork for your PRs, and not your fork's `master` branch. This lets you have multiple PRs open and avoids issues if you mistakenly rebase a branch on your fork's `master` branch instead of the upstream `master` branch.

File: drivers/gles3/storage/texture_storage.cpp
Comment (clayjohn) [inline drivers/gles3/storage/texture_storage.cpp:1652]: We shouldn't generate mipmaps unless we need them

File: drivers/gles3/storage/texture_storage.h
Comment (clayjohn) [inline drivers/gles3/storage/texture_storage.h:322]: This is unused and so should be deleted

File: drivers/gles3/rasterizer_scene_gles3.cpp
Comment (clayjohn) [inline drivers/gles3/rasterizer_scene_gles3.cpp:1991]: This logic should only run if needed. It should not run every frame as that will needlessly hurt performance for users who dont use the screen or depth textures

**Review by clayjohn — COMMENTED** [review]
I don't think we should be always allocating a new buffer just to use for copying the screen. The buffer should only be allocated when needed (and should only have mipmaps if needed).

We already have this logic for 2D using the backbuffer. The backbuffer that can be optionally created on demand with ``_create_render_target_backbuffer()``

Since 2D and 3D share a rendertarget in the gl_compatibility renderer, they can also share a backbuffer. 

What I would do is as follows:
1. Use the backbuffer instead of creating a new framebuffer
2. Add support for a depth texture in the backbuffer
3. Only allocate the backbuffer when screen texture or depth texture are requested by a shader
4. Only copy the screen texture or depth texture when they will be used by a shader

**Comment by the-brickster** [comment]
@clayjohn let me know if there any more changes that need to be done to line up with the spec

**Comment by the-brickster** [comment]
I believe the extra line breaks should be gone, it's just remnants of removing my testing code

**Comment by akien-mga** [comment]
There's some trailing whitespace in this file: https://github.com/godotengine/godot/actions/runs/4094173769/jobs/7068178311

Could you squash the commits? See [PR workflow](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase) for instructions.

**Comment by fire** [comment]
I did a rebase to godotengine/godot if you need help.

https://github.com/V-Sekai/godot/tree/gles3-screen-depth

**Comment by the-brickster** [comment]
@fire Thanks so much for the help

File: drivers/gles3/storage/texture_storage.cpp
Comment (BastiaanOlij) [inline drivers/gles3/storage/texture_storage.cpp:1901]: As this could be either color buffer, or depth buffer, or both, we probably need to check the value before we delete.

**Review by BastiaanOlij — APPROVED** [review]
This looks good to me,

Would be good for @dsnopek and @clayjohn to have a quick look as well.

**Comment by dsnopek** [comment]
This seems good to me (aside from Bastiaan's code comment above that I seconded), but I don't feel confident enough in my rendering skills to add an approval. If Clay can take a look at it, that would be ideal!

File: drivers/gles3/storage/texture_storage.cpp
Comment (clayjohn) [inline drivers/gles3/storage/texture_storage.cpp:1797]: If using multiview, I don't think the backbuffer can be shared with 2D as in 2D the backbuffer is always assumed to be a Texture_2D. We may need to create a separate backbuffer for multiview

@dsnopek How does 2D rendering work in XR when using Multiview? Does it just render to the first view index, or does it require using both views? From my brief look at the shader code and canvas render code it looks like 2D rendering should still just take place on a texture2D

**Review by clayjohn — CHANGES_REQUESTED** [review]

File: drivers/gles3/storage/texture_storage.cpp
Comment (dsnopek) [reply drivers/gles3/storage/texture_storage.cpp:1797]: > How does 2D rendering work in XR when using Multiview?

That's a good question! We probably want to just disable 2D rendering on multiview XR viewports, but I'll need to test to see what is actually happening now.

(We've got to be careful to NOT disable 2D rendering for non-multiview XR apps, since smartphone AR still needs to be able to render 2D stuff.)

The problem is that (a) rendering to just the first view, or (b) rendering the exact same thing to all views, are both wrong. :-) We'd need to render something slightly different to each view to get the right stereoscopic effect.

For the most part, stereo XR apps usually render 2D stuff onto a mesh in the 3D world, rather than rendering directly on to the output FBO.

Comment (dsnopek) [reply drivers/gles3/storage/texture_storage.cpp:1797]: I just tried putting a CanvasLayer with a ColorRect on it into a VR test app with gl_compatibility, and it didn't show up at all!

File: drivers/gles3/storage/texture_storage.cpp
Comment (the-brickster) [reply drivers/gles3/storage/texture_storage.cpp]: Are the changes in the recent commit appropriate?

File: drivers/gles3/storage/texture_storage.cpp
Comment (the-brickster) [reply drivers/gles3/storage/texture_storage.cpp:1797]: is there a way I could test without needing a VR headset?

Comment (dsnopek) [reply drivers/gles3/storage/texture_storage.cpp:1797]: > is there a way I could test without needing a VR headset?

On Linux, you can use Monado to have a simulated OpenXR headset, which you control with keyboard and mouse. I'm not sure what's possible on other platforms.

Or, you can also use the `MobileVRInterface`, which just renders the two views to the screen, without needing a headset.

BTW, my last comment about the ColorRect not showing was a good thing! We don't want 2D stuff to render on top of 3D with stereoscopic XR.

Comment (clayjohn) [reply drivers/gles3/storage/texture_storage.cpp:1797]: For the purposes of this PR I guess we can assume that if ``use_multiview`` is true, we can just forget about 2D rendering. 

Therefore, in this section when ``use_multiview`` is false we can call ``_create_render_target_backbuffer`` to create the backbuffer with full mipmaps, so it can be shared with 2D

File: drivers/gles3/storage/texture_storage.cpp
Comment (the-brickster) [reply drivers/gles3/storage/texture_storage.cpp]: Added a separate check if the backbuffer depth is defined in the delete section

File: drivers/gles3/storage/texture_storage.cpp
Comment (clayjohn) [inline drivers/gles3/storage/texture_storage.cpp:1791]:
```suggestion
	if (rt->backbuffer_fbo == 0) {
		glGenFramebuffers(1, &rt->backbuffer_fbo);
	}
```
We need to check if the framebuffer already exists. If it does, we do not want to generate a new one and overwrite the old one.

File: drivers/gles3/rasterizer_scene_gles3.cpp
Comment (clayjohn) [inline drivers/gles3/rasterizer_scene_gles3.cpp:1998]:
```suggestion
		texture_storage->copy_scene_to_backbuffer(rt, scene_state.used_screen_texture, scene_state.used_depth_texture);
		glBindFramebuffer(GL_READ_FRAMEBUFFER, rt->fbo);
		glReadBuffer(GL_COLOR_ATTACHMENT0);
```
We need to specify the READ_FRAMEBUFFER or else it can try to read from the last buffer specified as the READ_FRAMEBUFFER

**Review by clayjohn — CHANGES_REQUESTED** [review]
Running the code locally did not work with this current version. I had to make some changes myself to get it working. I have added those changes as suggestions. 

Once those changes are applied and this PR is rebased it should be ready to go (and can be merged shortly after we release Godot 4.0).

**Comment by the-brickster** [comment]
@clayjohn changes pushed and rebased against master

**Review by clayjohn — APPROVED** [review]
This looks good to go now. We can go ahead and merge it shortly after releasing 4.0 with an aim to get it into 4.1 and perhaps cherry pick to 4.0.1

**Pull request merged**
